### PR TITLE
Adding a fetcher for the functional surface data used in the surface searchlight example

### PR DESCRIPTION
**Pull request by SylvainTakerkart**
Hi guys, as planned during the sprint in April, here is a PR to create a fetcher for the functional data that is used in the example of the surface-based searchlight...

The data fetched here is a set of single-trial beta maps from one subject of the InterTVA dataset (https://openneuro.org/datasets/ds001771/), projected on the fsaverage5 surface.

### CODE REVIEW

**Comment by SylvainTakerkart** [comment]
@kchawla-pi  : for now, the data is hosted on our owncloud in Marseille... Could you download the zip file from  https://cloud.int.univ-amu.fr/index.php/s/29fJPH5Gm2H9Pxj/download , upload it on OSF and give me the new URL so that I can change it in the code?
Thanks,
Sylvain

**Comment by kchawla-pi** [comment]
Hey Sylvain, the data has been uploaded here: https://osf.io/aekp7/

**Comment by kchawla-pi** [comment]
@SylvainTakerkart We are aiming for an end of September release. Think this can be included?

**Comment by SylvainTakerkart** [comment]
trying to get back into it...
@kchawla-pi: I realized there was a small mistake in the data file (some redundancy with some extra files not needed...): can you re-download the zip file here ( https://cloud.int.univ-amu.fr/index.php/s/mxAj4M5XfzTGoBN ) and replace the one that you put on osf with this new one?
thanks!
sylvain

**Comment by kchawla-pi** [comment]
Done. Please confirm the download has been correctly updated.

**Comment by SylvainTakerkart** [comment]
the new file and its content look ok, thanks! but it seems I still have a problem with the use of _fetch_files... maybe we can talk about it when you're in Marseille next week...

**Comment by kchawla-pi** [comment]
Hi @SylvainTakerkart , I just ran the fetch_surf_tva_localizer function & I don't see a duplication problem. There're only the uncompressed files, not the compressed one.
Would you delete the existing material pertaining to this dataset & run it again?

**Comment by kchawla-pi** [comment]
`test_fetch_surf_tva_localizer` is failing cuz `fetch_surf_tva_localizer` returns a dict with 4 keys not 3. It would be better to assert the keys present than their number.

**Comment by kchawla-pi** [comment]
@GaelVaroquaux In response to https://github.com/nilearn/nilearn/pull/1959#pullrequestreview-331236802 , I think you can review this PR, we can merge this into that one.

WDYT?

**Comment by kchawla-pi** [comment]
@SylvainTakerkart You have the file examples/02_decoding/plot_intertva_surface_searchlight.py in both PRs. This PR is only about incorporating the fetcher, not writing examples with it.
If there are difference between them, incorporate any desirable changes into the file in PR #1959  and remove it from here.
This will also ensure that this PR will no longer have the unresolvable CircleCI failures, and make it easier to review.
Once this is merged, you can work on that.
Pull from master for both PRs.

File: nilearn/datasets/func.py [hunk 1]
```diff
@@ -2126,6 +2126,75 @@ def fetch_development_fmri(n_subjects=None, reduce_confounds=True,
...
+        Path of the data directory. Used to force data storage in a specified
+        location. Default: None
+
+    verbose: int, optional (default 1)
+        Defines the level of verbosity of the output.
+
+    Returns
+    -------
+    data: sklearn.datasets.base.Bunch
+        Dictionary-like object, the interest attributes are :
```
Comment (kchawla-pi) [inline nilearn/datasets/func.py]: Replace `interest` with `useful`.

File: nilearn/datasets/description/intertva_localizer_surface.rst [hunk 2]
```diff
@@ -0,0 +1,28 @@
+InterTVA localizer surface
+
+
+Notes
+-----
+A subset of the InterTVA data set, dedicated to run the example of the cortex-based surfacic searchlight of nilearn.
```
Comment (kchawla-pi) [inline nilearn/datasets/description/intertva_localizer_surface.rst]: surfacic?

File: nilearn/datasets/func.py [hunk 3]
```diff
@@ -2126,6 +2126,75 @@ def fetch_development_fmri(n_subjects=None, reduce_confounds=True,
...
+    # Preliminary checks and declarations
+    dataset_name = 'intertva_localizer_surface'
+    intertva_dir = _get_dataset_dir(dataset_name, data_dir=data_dir)
+    dataset_dir = _get_dataset_dir('InterTVA_sub-41_surfdata', data_dir=intertva_dir)
+    url = 'https://osf.io/aekp7/download'
+
+    # Dataset description
+    fdescr = _get_dataset_descr(dataset_name)
+
+    # First, get the metadata, i.e the labels to be used for the searchlight decoding
```
Comment (kchawla-pi) [inline nilearn/datasets/func.py]: Unnecessary comment, the code is well written to explain what's happening here.

File: nilearn/datasets/func.py [hunk 4]
```diff
@@ -2126,6 +2126,75 @@ def fetch_development_fmri(n_subjects=None, reduce_confounds=True,
...
+    label_path = 'labels_voicelocalizer_voice_vs_nonvoice.tsv'
+    label_file = (label_path, url , {'uncompress': True})
+
+    label_localpath = _fetch_files(dataset_dir, [label_file], resume=resume,verbose=verbose)[0]
+    labels = np.genfromtxt(label_localpath, dtype='str', skip_header=1)[:,0]
+
+    # Secondly, get the mask of a region of interest for this task
```
Comment (kchawla-pi) [inline nilearn/datasets/func.py]: Unnecessary comment, the code is well written to explain what's happening here.

File: nilearn/datasets/func.py [hunk 5]
```diff
@@ -2126,6 +2126,75 @@ def fetch_development_fmri(n_subjects=None, reduce_confounds=True,
...
+    mask_path = 'fsaverage5.lh.stg_sts.gii'
+    mask_file = (mask_path, url, {'uncompress': True})
+    mask = _fetch_files(dataset_dir, [mask_file], resume=resume, verbose=verbose)[0]
+
+    # Finally, get the data itself, i.e the 144 single-trial beta maps projected on the surface
+    gifti_file_list = []
+    for current_trial in range(144):
+        gifti_path = 'fsaverage5.lh.beta_{:04d}.gii'.format(current_trial+1)
+        gifti_file = (gifti_path, url, {'uncompress': True})
```
Comment (kchawla-pi) [inline nilearn/datasets/func.py]: Replace the name `gifti_file` with something like `download_info_gifti_`, since this is not in-fact a gifti-file.

File: nilearn/datasets/func.py [hunk 6]
```diff
@@ -2126,6 +2126,75 @@ def fetch_development_fmri(n_subjects=None, reduce_confounds=True,
...
+    label_localpath = _fetch_files(dataset_dir, [label_file], resume=resume,verbose=verbose)[0]
+    labels = np.genfromtxt(label_localpath, dtype='str', skip_header=1)[:,0]
+
+    # Secondly, get the mask of a region of interest for this task
+    mask_path = 'fsaverage5.lh.stg_sts.gii'
+    mask_file = (mask_path, url, {'uncompress': True})
+    mask = _fetch_files(dataset_dir, [mask_file], resume=resume, verbose=verbose)[0]
+
+    # Finally, get the data itself, i.e the 144 single-trial beta maps projected on the surface
+    gifti_file_list = []
```
Comment (kchawla-pi) [inline nilearn/datasets/func.py]: Rename this to `download_info_giftis`

File: nilearn/datasets/tests/test_func.py [hunk 7]
```diff
@@ -679,3 +679,17 @@ def test_fetch_development_fmri_exception():
     with pytest.raises(ValueError, match='Wrong value for age_group'):
         func._filter_func_regressors_by_participants(participants='junk',
                                                      age_group='junk for test')
+
+def test_fetch_surf_tva_localizer():
+    # get the data
```
Comment (kchawla-pi) [inline nilearn/datasets/tests/test_func.py]: Unnecessary comment.

File: nilearn/datasets/tests/test_func.py [hunk 8]
```diff
@@ -679,3 +679,17 @@ def test_fetch_development_fmri_exception():
     with pytest.raises(ValueError, match='Wrong value for age_group'):
         func._filter_func_regressors_by_participants(participants='junk',
                                                      age_group='junk for test')
+
+def test_fetch_surf_tva_localizer():
```
Comment (kchawla-pi) [inline nilearn/datasets/tests/test_func.py]: We must create a better test. Downloading all the files is a long process, and is fragile if the server is interrupted.

File: nilearn/datasets/description/intertva_localizer_surface.rst (hunk 2 above)
Comment (bthirion) [reply nilearn/datasets/description/intertva_localizer_surface.rst]: surface-based

File: nilearn/datasets/func.py (hunk 4 above)
Comment (bthirion) [reply nilearn/datasets/func.py]: Would be useful to state that the regions is 'superior temporal gyrus  + superior temporal sulcus'

File: nilearn/datasets/func.py [hunk 9]
```diff
@@ -2126,6 +2126,75 @@ def fetch_development_fmri(n_subjects=None, reduce_confounds=True,
...
+        Defines the level of verbosity of the output.
+
+    Returns
+    -------
+    data: sklearn.datasets.base.Bunch
+        Dictionary-like object, the interest attributes are :
+         - 'func_left': Paths to Gifti files containing resting state
+                        time series left hemisphere
+         - 'phenotypic': array containing tuple with subject ID, age,
+                         dominant hand and sex for each subject.
```
Comment (bthirion) [inline nilearn/datasets/func.py]: IIUC, it mostly (uniquerly ?) constains the labels information

File: nilearn/datasets/func.py (hunk 1 above)
Comment (SylvainTakerkart) [reply nilearn/datasets/func.py]: @kchawla-pi ; sorry, this looks like a bad copy-paste on my part... from what I've just checked, "interest" is used in the same sentence in several other fetcher functions of functional data (in the same func.py file)... shall I fix all of them (i.e replace with "useful") in the present PR?

File: nilearn/datasets/func.py (hunk 9 above)
Comment (SylvainTakerkart) [reply nilearn/datasets/func.py]: done! (sorry, there was some bad copy-paste on my part...)

File: nilearn/datasets/tests/test_func.py (hunk 8 above)
Comment (SylvainTakerkart) [reply nilearn/datasets/tests/test_func.py]: Any idea? I have to admit I copied the content of the test function on some other existing ones... Isn't this a more global question (more global than what is done in this PR)?

Other than this, I think I fixed everything else...

File: nilearn/datasets/func.py (hunk 1 above)
Comment (kchawla-pi) [reply nilearn/datasets/func.py]: No, just your code. Overreaching in a PR makes it harder to review.

File: nilearn/datasets/tests/test_func.py (hunk 8 above)
Comment (kchawla-pi) [reply nilearn/datasets/tests/test_func.py]: Let me think on this, you can work on BT's comments.

**Comment by SylvainTakerkart** [comment]
Hi!
Getting back to this...
Everything seems to be ok, the CercleCI tests seem to have been successfull, but this does not show on github... @kchawla-pi, any clue why?
Sylvain

File: nilearn/datasets/func.py (hunk 1 above)
Comment (SylvainTakerkart) [reply nilearn/datasets/func.py]: Actually, this has been dealt with a while ago...

**Comment by kchawla-pi** [comment]
I can see the CircleCI entry here on github. can you not?

It can happen if the PR author is committing to the fork and also have their own CircleCI account. THen instead of showing up here, the CircleCI entry only shows up in their account on CircleCI.com/dashboard

**Comment by SylvainTakerkart** [comment]
I see the CircleCI entry here... but it's been saying "pending" for more than 24h now, whereas when you go to the CircleCI site by clicking on details, it says it ended successfully a day ago... anything I did wrong?

**Comment by kchawla-pi** [comment]
I believe that example is in a separate PR

**Comment by emdupre** [comment]
> I see the CircleCI entry here... but it's been saying "pending" for more than 24h now, whereas when you go to the CircleCI site by clicking on details, it says it ended successfully a day ago... anything I did wrong?

It looks like the error was on GitHub's side -- there's a notification at the top of Circle saying they tried to send back the status message but got a 500 response. Should be fixed on the next commit !

**Comment by SylvainTakerkart** [comment]
@bthirion @kchawla-pi 
yes, this dataset is for the example provided in in PR #1959...

**Comment by kchawla-pi** [comment]
@GaelVaroquaux @bthirion I don't have anything further to add.
I believe this still needs addressing, https://github.com/nilearn/nilearn/pull/2086#discussion_r357947744 , I am not sure how the URL mocker works, so can you chip into this and address that?


Once this is ready, we can merge this PR into #1959 (whcih must also be reviewed before we can merge this)

**Comment by bthirion** [comment]
To me the only pending issue is the one raised by @kchawla-pi  on testing. 
Then merging into #1959 makes sense of course.
Thx for keeping this running !

**Comment by SylvainTakerkart** [comment]
> To me the only pending issue is the one raised by @kchawla-pi on testing.

@bthirion, @emdupre , @kchawla-pi , any hint on how to improve testing? originally, I'd simply "copied" my testing function from other similar fetchers; I'd seen  @kchawla-pi comment on testing a while ago and asked for suggestions, but havn't got any at this point, and I'm not an expert...

can you suggest another fetcher from which I should get inspiration? (I mean, better inspiration ;) )

or maybe "improving testing" is a more general matter for all fetchers than just for this PR? WDYT?

**Comment by bthirion** [comment]
I think that the point is to have a mocking mechanism as done for the localizer, to avoid actually downloading the data during the tests.

**Comment by SylvainTakerkart** [comment]
Hi!
Just wondering whether we should give up on this PR? This new dataset was intendend to demonstrate the SurfaceSearchlight, but the recent example of #2435 takes another option... Here is a summary of the pros and cons from my point of view:
- following the example of #2435, we do not need a new dataset, but the example runs fairly slowly (1.fetch the full haxby raw data for one subject, which is fairly heavy; 2. project it onto the surface; 3. run the surface searchlight)
- using this dataset, the same example would run much faster (5MB download instead of almost 300MB; and no need to project to the surface), but it's an extra dataset...
Let me know!

**Comment by bthirion** [comment]
Given the tough context with datasets, I'm in favor of not adding another one.

**Comment by SylvainTakerkart** [comment]
ok, closing then...
